File: exercism/haskell/grains/Grains.hs
module Grains
  ( square
  , total )
where

-- square 1 = 1
-- square n = 2 * square (n-1)
square :: Int -> Integer
square x = 2^(x-1)

-- total = sum $ map (2^) [1..64]
--       = 2^0 + (2^1 + ... + 2^63) - 2^0
--       = 2^64 - 1
total :: Integer
total = 2^(64 :: Int) - 1

File: exercism/new/go/grains/grains.go
package grains

import (
	"errors"
)

// Total returns the total number of grains on the chessboard.
func Total() uint64 {
	// 2^0 + 2^1 + ... + 2^63 = 2^64 - 1
	return 1<<64 - 1
}

// Square returns the number of grains on a given square (1-based).
func Square(n int) (uint64, error) {
	if n < 1 || n > 64 {
		return 0, errors.New("Input value not in domain")
	}
	return 1 << (n - 1), nil
}
